File: fixtures/nodejs/j13/tests/password-notification.test.js
// Acceptance criteria tests for password expiry notification feature
const { User, AuthService, EmailService } = require('../fixture');

// The model must provide a PasswordNotificationService (or similar) in password-notification.js
let PasswordNotificationService;
try {
  const mod = require('./password-notification');
  PasswordNotificationService = mod.PasswordNotificationService || mod.default;
} catch (e) {
  // Will fail gracefully
}

describe('Password Expiry Notification - Acceptance Criteria', () => {
  let emailService;
  let userRepository;
  let users;

  beforeEach(() => {
    const ninetyOneDaysAgo = new Date(Date.now() - 91 * 24 * 60 * 60 * 1000);
    const tenDaysAgo = new Date(Date.now() - 10 * 24 * 60 * 60 * 1000);

    users = [
      new User({ id: '1', email: 'expired@test.com', name: 'Expired User', lastPasswordChange: ninetyOneDaysAgo }),
      new User({ id: '2', email: 'fresh@test.com', name: 'Fresh User', lastPasswordChange: tenDaysAgo }),
      new User({ id: '3', email: 'notified@test.com', name: 'Already Notified', lastPasswordChange: ninetyOneDaysAgo, passwordNotificationSentAt: new Date() }),
    ];

    userRepository = {
      findAll: jest.fn().mockResolvedValue(users),
      update: jest.fn().mockImplementation((id, data) => {
        const user = users.find(u => u.id === id);
        if (user) Object.assign(user, data);
        return Promise.resolve(user);
      }),
    };

    emailService = new EmailService();
    jest.spyOn(emailService, 'sendPasswordExpiryNotification');
  });

  test('notifies user whose password is 90+ days old', async () => {
    if (!PasswordNotificationService) {
      console.warn('PasswordNotificationService not found — test skipped');
      return;
    }
    const authService = new AuthService(userRepository);
    const service = new PasswordNotificationService(authService, emailService);
    await service.run();
    expect(emailService.sendPasswordExpiryNotification).toHaveBeenCalledWith(
      expect.objectContaining({ email: 'expired@test.com' })
    );
  });

  test('does not notify user whose password is recent', async () => {
    if (!PasswordNotificationService) return;
    const authService = new AuthService(userRepository);
    const service = new PasswordNotificationService(authService, emailService);
    await service.run();
    expect(emailService.sendPasswordExpiryNotification).not.toHaveBeenCalledWith(
      expect.objectContaining({ email: 'fresh@test.com' })
    );
  });

  test('does not send duplicate notifications', async () => {
    if (!PasswordNotificationService) return;
    const authService = new AuthService(userRepository);
    const service = new PasswordNotificationService(authService, emailService);
    await service.run();
    expect(emailService.sendPasswordExpiryNotification).not.toHaveBeenCalledWith(
      expect.objectContaining({ email: 'notified@test.com' })
    );
  });

  test('records notification sent timestamp', async () => {
    if (!PasswordNotificationService) return;
    const authService = new AuthService(userRepository);
    const service = new PasswordNotificationService(authService, emailService);
    await service.run();
    expect(userRepository.update).toHaveBeenCalledWith(
      '1',
      expect.objectContaining({ passwordNotificationSentAt: expect.any(Date) })
    );
  });
});

File: fixtures/java/j13/tests/src/test/java/PasswordNotificationTest.java
// Acceptance criteria tests for password expiry notification feature
// Functionally equivalent to the Node.js password-notification.test.js

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Arrays;
import java.util.List;

class PasswordNotificationTest {

    private EmailService emailService;
    private UserRepository userRepository;
    private List<User> users;

    @BeforeEach
    void setUp() {
        Instant ninetyOneDaysAgo = Instant.now().minus(91, ChronoUnit.DAYS);
        Instant tenDaysAgo = Instant.now().minus(10, ChronoUnit.DAYS);

        users = Arrays.asList(
            new User("1", "expired@test.com", "Expired User", ninetyOneDaysAgo, null),
            new User("2", "fresh@test.com", "Fresh User", tenDaysAgo, null),
            new User("3", "notified@test.com", "Already Notified", ninetyOneDaysAgo, Instant.now())
        );

        userRepository = mock(UserRepository.class);
        when(userRepository.findAll()).thenReturn(users);
        when(userRepository.update(anyString(), any(User.class))).thenAnswer(invocation -> {
            String userId = invocation.getArgument(0);
            User data = invocation.getArgument(1);
            User user = users.stream().filter(u -> u.getId().equals(userId)).findFirst().orElse(null);
            if (user != null && data.getPasswordNotificationSentAt() != null) {
                user.setPasswordNotificationSentAt(data.getPasswordNotificationSentAt());
            }
            return user;
        });

        emailService = spy(new EmailService());
    }

    @Test
    void notifiesUserWhosePasswordIs90PlusDaysOld() {
        AuthService authService = new AuthService(userRepository);
        PasswordNotificationService service = new PasswordNotificationService(authService, emailService);
        service.run();

        verify(emailService).sendPasswordExpiryNotification(
            argThat(user -> "expired@test.com".equals(user.getEmail()))
        );
    }

    @Test
    void doesNotNotifyUserWhosePasswordIsRecent() {
        AuthService authService = new AuthService(userRepository);
        PasswordNotificationService service = new PasswordNotificationService(authService, emailService);
        service.run();

        verify(emailService, never()).sendPasswordExpiryNotification(
            argThat(user -> "fresh@test.com".equals(user.getEmail()))
        );
    }

    @Test
    void doesNotSendDuplicateNotifications() {
        AuthService authService = new AuthService(userRepository);
        PasswordNotificationService service = new PasswordNotificationService(authService, emailService);
        service.run();

        verify(emailService, never()).sendPasswordExpiryNotification(
            argThat(user -> "notified@test.com".equals(user.getEmail()))
        );
    }

    @Test
    void recordsNotificationSentTimestamp() {
        AuthService authService = new AuthService(userRepository);
        PasswordNotificationService service = new PasswordNotificationService(authService, emailService);
        service.run();

        verify(userRepository).update(
            eq("1"),
            argThat(user -> user.getPasswordNotificationSentAt() != null)
        );
    }
}
